http-common.c

```diff
@@ -113,36 +113,157 @@ set_method(CURL *ehandle, enum http_method method, struct sized_buffer *body)
+    curl_easy_setopt(ehandle, CURLOPT_POSTFIELDSIZE, body->size);
+  }
 }
 
 void
-set_url(CURL *ehandle, char *base_api_url, char endpoint[])
+set_url(CURL *ehandle, char base_api_url[], char endpoint[], va_list *args)
 {
+  //create the url route
+  char url_route[MAX_URL_LEN];
+  int ret = vsnprintf(url_route, sizeof(url_route), endpoint, *args);
+  ASSERT_S(ret < (int)sizeof(url_route), "oob write of url_route");
+
   char base_url[MAX_URL_LEN];
-  int ret = snprintf(base_url, sizeof(base_url), "%s%s", base_api_url, endpoint);
+  ret = snprintf(base_url, sizeof(base_url), "%s%s", base_api_url, url_route);
   ASSERT_S(ret < (int)sizeof(base_url), "Out of bounds write attempt");
 
   CURLcode ecode = curl_easy_setopt(ehandle, CURLOPT_URL, base_url);
   ASSERT_S(CURLE_OK == ecode, curl_easy_strerror(ecode));
 }
 
+static void
+default_cb(void *data)
+{ 
+  return; 
+  (void)data; 
+}
+
+static perform_action
+default_success_cb(
+  void *p_data,
+  enum http_code code,
+  struct sized_buffer *body,
+  struct api_header_s *pairs)
+{ 
+  return ACTION_SUCCESS;
+  (void)p_data; 
+  (void)code; 
+  (void)body;
+  (void)pairs;
+}
+
+static perform_action
+default_retry_cb(
+  void *p_data,
+  enum http_code code,
+  struct sized_buffer *body,
+  struct api_header_s *pairs)
+{
+  return ACTION_RETRY;
+  (void)p_data; 
+  (void)code;
+  (void)body;
+  (void)pairs;
+}
+
+static perform_action 
+default_abort_cb(
+  void *p_data,
+  enum http_code code,
+  struct sized_buffer *body,
+  struct api_header_s *pairs)
+{ 
+  return ACTION_ABORT; 
+  (void)p_data; 
+  (void)code; 
+  (void)body;
+  (void)pairs;
+}
+
+void
+perform_request(
+  struct sized_buffer *body,
+  struct api_header_s *pairs,
+  CURL *ehandle,
+  struct perform_cbs *cbs)
+{
+  if (!cbs->start) cbs->start = &default_cb;
+  if (!cbs->before_perform) cbs->before_perform = &default_cb;
+  if (!cbs->on_1xx) cbs->on_1xx = &default_success_cb;
+  if (!cbs->on_2xx) cbs->on_2xx = &default_success_cb;
+  if (!cbs->on_3xx) cbs->on_3xx = &default_success_cb;
+  if (!cbs->on_4xx) cbs->on_4xx = &default_abort_cb;
+  if (!cbs->on_5xx) cbs->on_5xx = &default_retry_cb;
+
+
+  /* callback mean't to be used for setting stuff up 
+   *  before connections attempts */
+  (*cbs->start)(cbs->p_data);
+
+  perform_action action;
+  do {
+    /* triggers on every start of loop iteration */
+    (*cbs->before_perform)(cbs->p_data);
+  
+    CURLcode ecode;
+    //perform the connection
+    ecode = curl_easy_perform(ehandle);
+    ASSERT_S(CURLE_OK == ecode, curl_easy_strerror(ecode));
+
+    //get response's code
+    enum http_code code;
+    ecode = curl_easy_getinfo(ehandle, CURLINFO_RESPONSE_CODE, &code);
+    ASSERT_S(CURLE_OK == ecode, curl_easy_strerror(ecode));
+
+    //get request's url
+    const char *url = NULL;
+    ecode = curl_easy_getinfo(ehandle, CURLINFO_EFFECTIVE_URL, &url);
+    ASSERT_S(CURLE_OK == ecode, curl_easy_strerror(ecode));
+
+    D_PRINT("Request URL: %s", url);
+
```

Comment:
it would be great some well known error code are handled with 
```
switch(code) 
   case HTTP_NO_CONTENT:
        reason = "The request completed succesfully but returned no content.";
        action = DONE;
        break;
      case HTTP_NOT_MODIFIED:
        reason = "The entity was not modified (no action was taken).";
        action = DONE;
        break;
      case HTTP_BAD_REQUEST:
        reason = "The request was improperly formatted, or the server couldn't understand it.";
        action = ABORT;
        break;
      case HTTP_UNAUTHORIZED:
        reason = "The Authorization header was missing or invalid.";
        action = ABORT;
        break;
      case HTTP_FORBIDDEN:
        reason = "The Authorization token you passed did not have permission to the resource.";
        action = DONE;
        break;
      case HTTP_NOT_FOUND:
        reason = "The resource at the location specified doesn't exist.";
        action = ABORT;
        break;
      case HTTP_METHOD_NOT_ALLOWED:
        reason = "The HTTP method used is not valid for the location specified.";
        action = ABORT;
        break;
      case HTTP_TOO_MANY_REQUESTS:
        cooldown;
      default:
         if (code >= 500) 
             action
         else if ( code >= 400)
             ...
```

Reply:
Btw, for the common "reasons" I added a function in http-common called http_reason_print